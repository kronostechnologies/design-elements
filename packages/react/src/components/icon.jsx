import React from 'react';

export default function() {
    return (
        <span></span>
    );
}
