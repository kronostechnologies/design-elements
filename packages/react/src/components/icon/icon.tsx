import AlertOctagon from 'feather-icons/dist/icons/alert-octagon.svg';
import AlertTriangle from 'feather-icons/dist/icons/alert-triangle.svg';
import ArrowDown from 'feather-icons/dist/icons/arrow-down.svg';
import ArrowLeft from 'feather-icons/dist/icons/arrow-left.svg';
import ArrowRight from 'feather-icons/dist/icons/arrow-right.svg';
import ArrowUp from 'feather-icons/dist/icons/arrow-up.svg';
import Calendar from 'feather-icons/dist/icons/calendar.svg';
import Check from 'feather-icons/dist/icons/check.svg';
import ChevronDown from 'feather-icons/dist/icons/chevron-down.svg';
import ChevronLeft from 'feather-icons/dist/icons/chevron-left.svg';
import ChevronRight from 'feather-icons/dist/icons/chevron-right.svg';
import ChevronUp from 'feather-icons/dist/icons/chevron-up.svg';
import ChevronsLeft from 'feather-icons/dist/icons/chevrons-left.svg';
import ChevronsRight from 'feather-icons/dist/icons/chevrons-right.svg';
import Copy from 'feather-icons/dist/icons/copy.svg';
import Edit from 'feather-icons/dist/icons/edit-2.svg';
import HelpCircle from 'feather-icons/dist/icons/help-circle.svg';
import Home from 'feather-icons/dist/icons/home.svg';
import Info from 'feather-icons/dist/icons/info.svg';
import Mail from 'feather-icons/dist/icons/mail.svg';
import MapPin from 'feather-icons/dist/icons/map-pin.svg';
import Menu from 'feather-icons/dist/icons/menu.svg';
import MoreHorizontal from 'feather-icons/dist/icons/more-horizontal.svg';
import MoreVertical from 'feather-icons/dist/icons/more-vertical.svg';
import Phone from 'feather-icons/dist/icons/phone.svg';
import PlusSign from 'feather-icons/dist/icons/plus.svg';
import Search from 'feather-icons/dist/icons/search.svg';
import Star from 'feather-icons/dist/icons/star.svg';
import Trash from 'feather-icons/dist/icons/trash.svg';
import X from 'feather-icons/dist/icons/x.svg';
import React, { VoidFunctionComponent } from 'react';
import Open from '../../icons/open.svg';
import Reorder from '../../icons/reorder.svg';

const iconMapping = {
    alertTriangle: AlertTriangle,
    alertOctagon: AlertOctagon,
    arrowDown: ArrowDown,
    arrowLeft: ArrowLeft,
    arrowRight: ArrowRight,
    arrowUp: ArrowUp,
    calendar: Calendar,
    check: Check,
    chevronDown: ChevronDown,
    chevronLeft: ChevronLeft,
    chevronRight: ChevronRight,
    chevronUp: ChevronUp,
    chevronsLeft: ChevronsLeft,
    chevronsRight: ChevronsRight,
    copy: Copy,
    edit: Edit,
    helpCircle: HelpCircle,
    home: Home,
    info: Info,
    mail: Mail,
    mapPin: MapPin,
    menu: Menu,
    moreHorizontal: MoreHorizontal,
    moreVertical: MoreVertical,
    open: Open,
    phone: Phone,
    plusSign: PlusSign,
    reorder: Reorder,
    search: Search,
    star: Star,
    trash: Trash,
    x: X,
} as const;

export type IconName = keyof typeof iconMapping;

interface IconProps {
    /** Name of the icon, has to be in IconName */
    name: IconName;
    /**
     * Size will affect both width and height
     * @default 24
     */
    size?: string;
    /**
     * Icon's color, in any format supported by css (keyword, rgb, hex, etc.)
     * @default currentColor
     */
    color?: string;
}

export const Icon: VoidFunctionComponent<IconProps> = ({
    name,
    size,
    color,
    ...props
}: IconProps) => {
    const Component = iconMapping[name];

    if (!Component) {
        return null;
    }

    return (
        <Component
            height={size}
            focusable={false}
            width={size}
            color={color}
            {...props /* eslint-disable-line react/jsx-props-no-spreading */}
        />
    );
};

Icon.defaultProps = {
    size: '24',
    color: 'currentColor',
};
