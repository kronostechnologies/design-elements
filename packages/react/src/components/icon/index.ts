export { Icon } from './icon';
export type { IconName, IconProps } from './icon';
