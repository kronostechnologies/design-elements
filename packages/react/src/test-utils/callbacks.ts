export function doNothing(): void {
    // do nothing
}
