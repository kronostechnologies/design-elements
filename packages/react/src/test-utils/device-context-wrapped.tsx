import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
