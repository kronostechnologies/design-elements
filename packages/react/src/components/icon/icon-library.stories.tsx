import React from 'react';
import { Icon } from './icon';

export default {
    title: 'Icon/library',
    component: Icon,
};

export const all = () => (
    <>
        <Icon name="alertTriangle" />
        <Icon name="arrowLeft" />
        <Icon name="check" />
        <Icon name="chevronDown" />
        <Icon name="chevronUp" />
        <Icon name="copy" />
        <Icon name="edit" />
        <Icon name="helpCircle" />
        <Icon name="home" />
        <Icon name="info" />
        <Icon name="mail" />
        <Icon name="mapPin" />
        <Icon name="menu" />
        <Icon name="moreHorizontal" />
        <Icon name="open" />
        <Icon name="phone" />
        <Icon name="search" />
        <Icon name="trash" />
        <Icon name="x" />
    </>
);

all.story = {
    decorators: [(storyFn: () => React.ReactElement) => (
        <div style={{ display: 'flex', justifyContent: 'space-around' }}>
            {storyFn()}
        </div>
    )],
};

export const alertTriangle = () => (
    <Icon name="alertTriangle" />
);

export const arrowLeft = () => (
    <Icon name="arrowLeft" />
);

export const check = () => (
    <Icon name="check" />
);

export const chevronDown = () => (
    <Icon name="chevronDown" />
);

export const chevronUp = () => (
    <Icon name="chevronUp" />
);

export const copy = () => (
    <Icon name="copy" />
);

export const edit = () => (
    <Icon name="edit" />
);

export const helpCircle = () => (
    <Icon name="helpCircle" />
);

export const home = () => (
    <Icon name="home" />
);

export const info = () => (
    <Icon name="info" />
);

export const mail = () => (
    <Icon name="mail" />
);

export const mapPin = () => (
    <Icon name="mapPin" />
);

export const menu = () => (
    <Icon name="menu" />
);

export const moreHorizontal = () => (
    <Icon name="moreHorizontal" />
);

export const open = () => (
    <Icon name="open" />
);

export const phone = () => (
    <Icon name="phone" />
);

export const search = () => (
    <Icon name="search" />
);

export const trash = () => (
    <Icon name="trash" />
);

export const x = () => (
    <Icon name="x" />
);
