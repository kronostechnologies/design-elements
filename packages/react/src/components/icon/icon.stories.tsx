import React from 'react';
import { Icon } from './icon';

export default {
    title: 'Icon',
    component: Icon,
};

export const icon = () => (
    <Icon name="alertTriangle" />
);

export const withColor = () => (
    <Icon name="alertTriangle" color="orange" />
);

export const withSize = () => (
    <Icon name="alertTriangle" size="78" />
);
