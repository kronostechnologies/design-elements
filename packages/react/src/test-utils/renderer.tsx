import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
