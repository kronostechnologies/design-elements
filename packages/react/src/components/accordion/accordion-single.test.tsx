import React from 'react';
import { ReactWrapper} from 'enzyme';
import { mountWithTheme} from '../../test-utils/renderer';
import { SingleOpenAccordionGroup, Accordion, AccordionItem } from './index';
import { Icon } from '../icon/icon';


describe('SingleOpenAccordionGroup renders with default state open', () => {

  let wrapper: ReactWrapper;

  beforeEach(() => {
    wrapper = mountWithTheme(
     
        <SingleOpenAccordionGroup defaultExpandedItemId="3">
          <Accordion title="Panel 1" id="1" noMargin>
            Content for Panel 1
          </Accordion>
          <Accordion title="Panel 2" id="2" noMargin>
            Content for Panel 2
          </Accordion>
          <Accordion title="Panel 3" id="3" noMargin>
            Content for Panel 3
          </Accordion>
        </SingleOpenAccordionGroup>
    );

  });

  it('matches snapshot (no margin)', () => {
    const tree = mountWithTheme(<Accordion title="Panel 1" id="1" noMargin>Content for Panel 1 </Accordion>);
    expect(tree).toMatchSnapshot();
  });


  it('should render Accordion components with expected props', () => {

    const accordions = wrapper.find(Accordion);

    // Ensure each Accordion component has the expected props
    accordions.forEach((element) => {
      const props = element.props();
      expect(props.id).toBeDefined();
      expect(props.title).toBeDefined();
      expect(props.children).toBeDefined();
    });
  });


  it('should only render Accordion components', () => {
    const wrapper =  mountWithTheme(
      <SingleOpenAccordionGroup>
        <Accordion id="1" title="Panel 1">
          Content for Panel 1
        </Accordion>
        <div>Invalid Child</div>
        <Accordion id="2" title="Panel 2">
          Content for Panel 2
        </Accordion>
      </SingleOpenAccordionGroup>
    );

    const accordions = wrapper.find(Accordion);

    // Ensure all elements in accordions list are of type Accordion
    accordions.forEach((element) => {
      expect(element.is(Accordion)).toBe(true);
    });

    // Ensure the invalid child is ignored (checking specific text content)
    const invalidChildText = wrapper.find('div').map((element) => element.text());
    expect(invalidChildText).not.toContain('Invalid Child');
  });


  it('should render with defaultExpandedItemId', () => {

      const firstAccordion = wrapper.find(Accordion).at(0);
      const secondAccordion = wrapper.find(Accordion).at(1);
      const thridAccordion = wrapper.find(Accordion).at(2);

      expect(firstAccordion.prop('isExpanded')).toBe(false);
      expect(secondAccordion.prop('isExpanded')).toBe(false);
      expect(thridAccordion.prop('isExpanded')).toBe(true);
  });

  it('should update expandedItemId state correctly when toggling to a different panel', () => {

    const setStateMock = jest.fn();
    const useStateMock: any = (initialValue: any) => [initialValue, setStateMock];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);

    const wrapper = mountWithTheme(
      <SingleOpenAccordionGroup defaultExpandedItemId="3">
        <Accordion title="Panel 1" id="1">
          Content for Panel 1
        </Accordion>
        <Accordion title="Panel 2" id="2">
          Content for Panel 2
        </Accordion>
        <Accordion title="Panel 3" id="3">
          Content for Panel 3
        </Accordion>
      </SingleOpenAccordionGroup>
    );

    const firstAccordionButton = wrapper.find(Accordion).at(0).find('button.accordion-button');
    firstAccordionButton.simulate('click');

    // Verify that the state update function was called with the expected value
    const updatedState = setStateMock.mock.calls[0][0];
    const expandedItemId = updatedState('3');
    expect(expandedItemId).toBe('1');

    jest.clearAllMocks();
  });

  it('should update expandedItemId state correctly when toggling to the same panel', () => {
    const setStateMock = jest.fn();
    const useStateMock: any = (initialValue: any) => [initialValue, setStateMock];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);

    const wrapper = mountWithTheme(
      <SingleOpenAccordionGroup defaultExpandedItemId="3">
        <Accordion title="Panel 1" id="1">
          Content for Panel 1
        </Accordion>
        <Accordion title="Panel 2" id="2">
          Content for Panel 2
        </Accordion>
        <Accordion title="Panel 3" id="3">
          Content for Panel 3
        </Accordion>
      </SingleOpenAccordionGroup>
    );

    const firstAccordionButton = wrapper.find(Accordion).at(0).find('button.accordion-button');
    firstAccordionButton.simulate('click');

    // Simulate another click on the same Accordion button
    firstAccordionButton.simulate('click');

    // Verify that the state update function was called with the expected value
    const updatedState = setStateMock.mock.calls[0][0];
    const expandedItemId = updatedState('1');
    expect(expandedItemId).toBe(null);

    jest.clearAllMocks();
  });

  it('renders an icon based on the isExpanded prop', () => {
    const wrapper = mountWithTheme(
      <AccordionItem
        headerId="header1"
        panelId="panel1"
        title="Accordion 1"
        type="medium"
        isExpanded={true}
      >
        Content for the accordion panel
      </AccordionItem>
    );

    const icon = wrapper.find(Icon);
    expect(icon.prop('name')).toBe('carretDown');
  });

  it('renders the opposite icon when isExpanded is false', () => {
    const wrapper = mountWithTheme(
      <AccordionItem
        headerId="header2"
        panelId="panel2"
        title="Accordion 2"
        type="medium"
        isExpanded={false}
      >
        Content for the accordion panel
      </AccordionItem>
    );

    const icon = wrapper.find(Icon);
    expect(icon.prop('name')).toBe('carretRight');
  });

});