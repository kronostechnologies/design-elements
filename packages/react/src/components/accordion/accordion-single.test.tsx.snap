// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleOpenAccordionGroup renders with default state open matches snapshot (no margin) 1`] = `
.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: inherit;
  border: 1px solid;
  border-radius: 1.5rem;
  box-sizing: border-box;
  color: inherit;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: inherit;
  font-size: 0.75rem;
  font-weight: var(--font-bold);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-letter-spacing: 0.025rem;
  -moz-letter-spacing: 0.025rem;
  -ms-letter-spacing: 0.025rem;
  letter-spacing: 0.025rem;
  line-height: 1rem;
  min-height: var(--size-2x);
  min-width: 2rem;
  outline: none;
  padding: 0 var(--spacing-2x);
  text-transform: uppercase;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4:focus {
  outline: none;
}

.c4:focus {
  outline: none;
  border-color: #006296;
  box-shadow: 0 0 0 2px #84C6EA;
}

.c4:not(:disabled) {
  cursor: pointer;
}

.c4 > svg {
  color: inherit;
  height: var(--size-1x);
  width: var(--size-1x);
}

.c3 {
  background-color: transparent;
  border-color: transparent;
  color: #60666E;
}

.c3:focus {
  outline: none;
}

.c3:focus {
  outline: none;
  border-color: #006296;
  box-shadow: 0 0 0 2px #84C6EA;
}

.c3:hover,
.c3[aria-expanded='true'] {
  background-color: #DBDEE1;
  color: #000000;
}

.c3:disabled {
  background-color: transparent;
  color: #B7BBC2;
}

.c3:focus {
  background-color: #FFFFFF;
  border-color: #006296;
  color: #60666E;
}

.c1 {
  font-size: 1.25rem;
  font-weight: var(--font-normal);
  line-height: 2rem;
  margin: 0;
}

.c5 {
  border: 0;
  border-top-width: 0;
  margin-bottom: 8px;
  border-radius: 0 0 4px 4px;
  background: #FAFAFA;
  color: #1B1C1E;
  font-size: 0.75rem;
  font-weight: 400;
  line-height: 1.7;
  -webkit-letter-spacing: 0.24px;
  -moz-letter-spacing: 0.24px;
  -ms-letter-spacing: 0.24px;
  letter-spacing: 0.24px;
  padding: 0;
  min-height: 0;
  overflow: hidden;
  height: 0;
  -webkit-transition: height 300ms cubic-bezier(0.25,0.46,0.45,0.94) 0ms;
  transition: height 300ms cubic-bezier(0.25,0.46,0.45,0.94) 0ms;
}

.c0 {
  border: 1px solid #60666E;
  border-bottom-color: #60666E;
  border-radius: 4px;
}

.c2 {
  text-transform: none;
  font-size: .875rem;
  line-height: 24px;
  font-weight: 400;
  -webkit-letter-spacing: 0.24px;
  -moz-letter-spacing: 0.24px;
  -ms-letter-spacing: 0.24px;
  letter-spacing: 0.24px;
  width: 100%;
  border-radius: 4px;
  padding: 8px;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
}

.c2[aria-expanded="true"] {
  background: #FFFFFF;
}

.c2 > svg {
  color: #60666E;
  padding: 4px;
  margin-right: 4px;
}

<Accordion
  id="1"
  noMargin={true}
  title="Panel 1"
>
  <AccordionItem
    headerId="idHeader1"
    noMargin={true}
    onToggle={[Function]}
    panelId="idPanel1"
    tag="h3"
    title="Panel 1"
    type="medium"
  >
    <Styled(Heading)
      className="accordion-header"
      isExpanded={false}
      noMargin={true}
      tag="h3"
      type="medium"
    >
      <Heading
        className="c0 accordion-header"
        isExpanded={false}
        noMargin={true}
        tag="h3"
        type="medium"
      >
        <styled.h3
          as="h3"
          className="c0 accordion-header"
          noMargin={true}
        >
          <h3
            className="c1 c0 accordion-header"
          >
            <Styled(Button)
              aria-controls="idPanel1"
              aria-expanded={false}
              buttonType="tertiary"
              className="accordion-button"
              id="idHeader1"
              isExpanded={false}
              label="Panel 1"
              onClick={[Function]}
            >
              <Button
                aria-controls="idPanel1"
                aria-expanded={false}
                buttonType="tertiary"
                className="c2 accordion-button"
                id="idHeader1"
                isExpanded={false}
                label="Panel 1"
                onClick={[Function]}
              >
                <Styled(AbstractButton)
                  aria-controls="idPanel1"
                  aria-expanded={false}
                  buttonType="tertiary"
                  className="c2 accordion-button"
                  id="idHeader1"
                  isExpanded={false}
                  isMobile={false}
                  onClick={[Function]}
                  type="button"
                >
                  <AbstractButton
                    aria-controls="idPanel1"
                    aria-expanded={false}
                    buttonType="tertiary"
                    className="c3 c2 accordion-button"
                    id="idHeader1"
                    isExpanded={false}
                    isMobile={false}
                    onClick={[Function]}
                    type="button"
                  >
                    <styled.button
                      aria-controls="idPanel1"
                      aria-expanded={false}
                      buttonType="tertiary"
                      className="c3 c2 accordion-button"
                      id="idHeader1"
                      isExpanded={false}
                      isMobile={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        aria-controls="idPanel1"
                        aria-expanded={false}
                        className="c4 c3 c2 accordion-button"
                        id="idHeader1"
                        onClick={[Function]}
                        type="button"
                      >
                        <Icon
                          aria-hidden="true"
                          color="currentColor"
                          name="carretRight"
                          size="24"
                        >
                          <svg
                            aria-hidden="true"
                            color="currentColor"
                            focusable={false}
                            height="24"
                            width="24"
                          />
                        </Icon>
                        Panel 1
                      </button>
                    </styled.button>
                  </AbstractButton>
                </Styled(AbstractButton)>
              </Button>
            </Styled(Button)>
          </h3>
        </styled.h3>
      </Heading>
    </Styled(Heading)>
    <styled.section
      aria-expanded={false}
      aria-labelledby="idHeader1"
      className="accordion-content"
      id="idPanel1"
      isExpanded={false}
      role="region"
    >
      <section
        aria-expanded={false}
        aria-labelledby="idHeader1"
        className="c5 accordion-content"
        id="idPanel1"
        role="region"
      >
        Content for Panel 1 
      </section>
    </styled.section>
  </AccordionItem>
</Accordion>
`;
